src/include/homestore/blk.h

```diff
@@ -24,122 +24,162 @@
-} __attribute__((__packed__));
-
-VENUM(BlkAllocStatus, uint32_t,
-      BLK_ALLOC_NONE = 0,       // No Action taken
-      SUCCESS = 1ul << 0,       // Success
-      FAILED = 1ul << 1,        // Failed to alloc/free
-      REQ_MORE = 1ul << 2,      // Indicate that we need more
-      SPACE_FULL = 1ul << 3,    // Space is full
-      INVALID_DEV = 1ul << 4,   // Invalid Device provided for alloc
-      PARTIAL = 1ul << 5,       // In case of multiple blks, only partial is alloced/freed
-      INVALID_THREAD = 1ul << 6 // Not possible to alloc in this thread
-);
-
-static_assert(sizeof(BlkId) < 8);
-#pragma pack(1)
-struct BlkId8_t : public BlkId {
-    uint8_t pad[8 - sizeof(BlkId)]{};
-
-    BlkId8_t& operator=(const BlkId& rhs) {
-        BlkId::operator=(rhs);
-        return *this;
-    }
+    bool operator==(MultiBlkId const& other) const { return (compare(*this, other) == 0); }
+    bool operator>(MultiBlkId const& other) const { return (compare(*this, other) > 0); }
+    bool operator<(MultiBlkId const& other) const { return (compare(*this, other) < 0); }
+
+    sisl::blob serialize();
+    uint32_t serialized_size() const;
+    void deserialize(sisl::blob const& b, bool copy);
+
+    bool has_room() const;
+    BlkId to_single_blkid() const;
+
+    static int compare(MultiBlkId const& one, MultiBlkId const& two);
+
+    struct iterator {
+        MultiBlkId const& mbid_;
+        uint16_t next_blk_{0};
+
+        iterator(MultiBlkId const& mb) : mbid_{mb} {}
+        std::optional< BlkId > next() {
+            if (next_blk_ == 0) {
+                auto bid = r_cast< BlkId const& >(mbid_);
+                ++next_blk_;
+                return (bid.is_valid()) ? std::make_optional(bid) : std::nullopt;
+            } else if (next_blk_ < mbid_.num_pieces()) {
+                auto cbid = mbid_.addln_pieces[next_blk_ - 1];
+                ++next_blk_;
+                return std::make_optional(BlkId{cbid.m_blk_num, cbid.m_nblks, mbid_.chunk_num()});
+            } else {
+                return std::nullopt;
+            }
+        }
+    };
+
+    iterator iterate() const;
 };
 #pragma pack()
-static_assert(sizeof(BlkId8_t) == 8);
 
-inline blk_num_t begin_of(const BlkId& blkid) { return blkid.get_blk_num(); }
-inline blk_num_t end_of(const BlkId& blkid) { return blkid.get_blk_num() + blkid.get_nblks(); }
-inline size_t hash_value(const BlkId& blkid) { return std::hash< uint64_t >()(blkid.to_integer()); }
 } // namespace homestore
 
-// hash function definitions
+///////////////////// hash function definitions /////////////////////
 namespace std {
 template <>
 struct hash< homestore::BlkId > {
-    typedef homestore::BlkId argument_type;
-    typedef size_t result_type;
-    result_type operator()(const argument_type& bid) const noexcept {
-        return std::hash< uint64_t >()(bid.to_integer());
+    size_t operator()(const homestore::BlkId& bid) const noexcept { return std::hash< uint64_t >()(bid.to_integer()); }
```

Comment:
nit: I wasn't aware of this utility. Won't `bid.to_integer()` served the purpose as hash key since it is unique between all BlkIds? Or there might be some cases that we want to store BlkIds from different SMs?

Reply:
Continous blkId is also one blkid which is still unique.

Reply:
sorry I mean
blkId{0, 0, 1} and blkid{0, 1, 1}   i.e chunk0, first blk and second blk,  in to_interger() they are continues which is 0/1.
hashing re-distribute them to the uint64 space, maybe 0 and 2048576, which re-distribute the continues blkID across whole uint64 space to avoid hot-spot.

Though honestly I dont see a use case in code right now

Reply:
std::hash of integer implementation is to just literal `no-op`, but we don't want to assume that. We have an integer and let the compiler/STL decides how to convert that to near unique. More over the to_integer() could return `uint32_t` and in that case, we need to make it `size_t`

Reply:
@xiaoxichen This is needed for us to put MultiBlkId in STL containers std::unordered_map, concurrent_hashmap etc

Reply:
@hkadayam we return uint64_t for to_integer(), not uint32_t, right? Isn't that good enough for `std::unordered_map<uint64_t /* bid.to_integer() as key */, xxx>` ?

Reply:
@yamingk Suggested practice for hash is to do std::hash(<integer>) even for integers. Let STL deal their evenness.